src/ai/backend/manager/models/kernel.py

```diff
@@ -620,15 +625,63 @@ async def set_kernel_status(
         if status in (KernelStatus.CANCELLED, KernelStatus.TERMINATED):
             data["terminated_at"] = now
 
-        async def _transit() -> None:
-            async with db.begin_session() as db_sess:
-                kernel_query = sa.select(KernelRow.status).where(KernelRow.id == kernel_id)
-                current_status = (await db_sess.execute(kernel_query)).scalar()
-                if status in KERNEL_STATUS_TRANSITION_MAP[current_status]:
-                    query = sa.update(KernelRow).values(**data).where(KernelRow.id == kernel_id)
-                    await db_sess.execute(query)
+        await cls.update_kernel(db, kernel_id, status, update_data=data)
+
+    @classmethod
+    async def update_kernel(
+        cls,
+        db: ExtendedAsyncSAEngine,
+        kernel_id: KernelId,
+        new_status: KernelStatus,
+        update_data: Optional[Mapping[str, Any]] = None,
+    ) -> bool:
+        """
+        Update kernel by given id and data.
+        Return True if the kernel is updated, else return False.
+        """
+
+        now = datetime.now(tzutc())
+
+        async def _update() -> bool:
+            async with db.begin_session() as db_session:
+                kernel_query = (
+                    sa.select(KernelRow)
+                    .where(KernelRow.id == kernel_id)
+                    .with_for_update()
+                    .options(
+                        noload("*"),
+                        load_only(KernelRow.status, KernelRow.session_id),
+                    )
+                )
+                kernel_row = (await db_session.scalars(kernel_query)).first()
+
+                if new_status not in KERNEL_STATUS_TRANSITION_MAP[kernel_row.status]:
+                    # TODO: log or raise error
+                    return False
```

Comment:
Evaluating validity of `new_status` before starting DB transaction will prevent creating useless transaction.
```suggestion
        if new_status not in KERNEL_STATUS_TRANSITION_MAP[kernel_row.status]:
            # TODO: log or raise error
            return False

        async def _update() -> bool:
            async with db.begin_session() as db_session:
                kernel_query = (
                    sa.select(KernelRow)
                    .where(KernelRow.id == kernel_id)
                    .with_for_update()
                    .options(
                        noload("*"),
                        load_only(KernelRow.status, KernelRow.session_id),
                    )
                )
                kernel_row = (await db_session.scalars(kernel_query)).first()
```

Reply:
we cannot validate the kernel status because the transaction fetches `kernel_row` by `kernel_id`.